src/goMain.ts

```diff
@@ -66,36 +66,45 @@ export function activate(ctx: vscode.ExtensionContext): void {
 	let useLangServer = vscode.workspace.getConfiguration('go')['useLanguageServer'];
 	setGlobalState(ctx.globalState);
 
-	updateGoPathGoRootFromConfig().then(() => {
+	updateGoPathGoRootFromConfig().then(async () => {
 		const updateToolsCmdText = 'Update tools';
-		const prevGoroot = ctx.globalState.get('goroot');
-		const currentGoroot = process.env['GOROOT'];
-		if (prevGoroot !== currentGoroot && prevGoroot) {
-			vscode.window.showInformationMessage('Your goroot is different than before, a few Go tools may need recompiling', updateToolsCmdText).then(selected => {
+		interface GoInfo {
+			goroot: string;
+			version: string;
+		}
+		const toolsGoInfo: { [id: string]: GoInfo; } = ctx.globalState.get('toolsGoInfo') || {};
+		const toolsGopath = getToolsGopath() || getCurrentGoPath();
+		if (!toolsGoInfo[toolsGopath]) {
+			toolsGoInfo[toolsGopath] = { goroot: null, version: null };
+		}
+		const prevGoroot = toolsGoInfo[toolsGopath].goroot;
+		const currentGoroot: string = process.env['GOROOT'];
+		if (prevGoroot && prevGoroot !== currentGoroot) {
+			vscode.window.showInformationMessage(`Your current goroot (${currentGoroot}) is different than before (${prevGoroot}), a few Go tools may need recompiling`, updateToolsCmdText).then(selected => {
 				if (selected === updateToolsCmdText) {
 					installAllTools(true);
 				}
 			});
 		} else {
-			getGoVersion().then(currentVersion => {
-				if (currentVersion) {
-					const prevVersion = ctx.globalState.get('goVersion');
-					const currVersionString = `${currentVersion.major}.${currentVersion.minor}`;
-
-					if (prevVersion !== currVersionString) {
-						if (prevVersion) {
-							vscode.window.showInformationMessage('Your Go version is different than before, few Go tools may need re-compiling', updateToolsCmdText).then(selected => {
-								if (selected === updateToolsCmdText) {
-									installAllTools(true);
-								}
-							});
-						}
-						ctx.globalState.update('goVersion', currVersionString);
+			const currentVersion = await getGoVersion();
+			if (currentVersion) {
+				const prevVersion = toolsGoInfo[toolsGopath].version;
```

Comment:
When this update goes out, `prevVersion` will be `null` for the first time, and users will get the prompt for recompiling.
We need a null check for `prevVersion` just like we do for `prevGoroot`

Reply:
yess, good catch

Reply:
we do, but just taht is nested , ill update to remove the nested if

Reply:
actually i think its fine how it is ? cuz we show the worning only if prev version exists

Reply:
Ah! I missed the nested check. looks good